plugins/janus_sip.c

```diff
@@ -1030,6 +1039,70 @@ static GHashTable *masters;
 static GHashTable *transfers;
 static janus_mutex sessions_mutex = JANUS_MUTEX_INITIALIZER;
 
+static void janus_event_init(janus_event_t *event)
+{
+	if (event == NULL)
+		return;
+
+	event->is_ready = FALSE;
+	janus_mutex_init(&event->mutex);
+	janus_condition_init(&event->cond);
+}
+
+static void janus_event_destroy(janus_event_t *event)
+{
+	if (event == NULL)
+		return;
+
+	event->is_ready = FALSE;
+	janus_condition_destroy(&event->cond);
+	janus_mutex_destroy(&event->mutex);
+}
+
+static void janus_event_reset(janus_event_t *event)
+{
+	if (event == NULL)
+		return;
+
+	event->is_ready = FALSE;
+}
+
+static void janus_event_set(janus_event_t *event)
+{
+	if (event == NULL)
+		return;
+
+	janus_mutex_lock(&event->mutex);
+	event->is_ready = TRUE;
+	janus_condition_signal(&event->cond);
+	janus_mutex_unlock(&event->mutex);
+}
+
+static gboolean janus_event_wait_timeout(janus_event_t *event, gint64 timeout)
+{
+	gboolean sts = FALSE;
+	gint64 end_time = 0;
+
+	if (event == NULL || timeout <= 0)
+		return sts;
+
+	janus_mutex_lock(&event->mutex);
+	end_time = g_get_monotonic_time() + timeout * G_TIME_SPAN_SECOND;
+	while (!event->is_ready)
+	{
+		sts = janus_condition_wait_until(&event->cond, &event->mutex, end_time);
+		if (!sts)
+			goto exit; // timeout occured
+	}
+
+	sts = TRUE;
+
+	exit:
+	janus_event_reset(event);
+	janus_mutex_unlock(&event->mutex);
+	return sts;
+}
```

Comment:
I think that this whole structure + functions are a bit overkill, if the only purpose is waiting for a string to be updated in response to the `nua_get_params`: this feels like something that could be done much more simply with the existing "wait" loop we have + maybe an atomic flag to trigger a state change.

Reply:
Yes, atomic flag + loop + sleep also can be used but I don't like "busy" wait.